All_home_works/hw2/task1.py

```diff
@@ -0,0 +1,96 @@
+"""
+Given a file containing text. Complete using only default collections:
+    1) Find 10 longest words consisting from largest amount of unique symbols
+    2) Find rarest symbol for document
+    3) Count every punctuation char
+    4) Count every non ascii char
+    5) Find most common non ascii char for document
+"""
+import string
+from collections import Counter, deque
+from typing import List
+
+TRANSLATE = str.maketrans({char: "" for char in string.punctuation})
+
+
+def _get_lines(file_path: str) -> str:
+    with open(file_path, encoding="raw_unicode_escape") as reading_file:
+        for line in reading_file.readlines():
+            line = line[:-1]
+            if line:
+                yield line
+
+
+def get_longest_diverse_words(file_path: str) -> List[str]:
+    queue = deque()
+    for line in _get_lines(file_path):
+        line = line.translate(TRANSLATE)
+        words = [(word, len(set(word))) for word in line.split() if word]
+        queue = _check_words(queue, words)
+
+    return [word[0] for word in reversed(queue)]
+
+
+def _check_words(queue, words):
+    for word, len_word in words:
+        if len(queue) < 10:
+            queue.append((word, len_word))
+            continue
+
+        queue = sorted(queue, key=lambda x: x[1])
```

Comment:
What is the algorithmic and memory complexity of this line?

Reply:
Sorted in python is a TimSort, so that means n*log(n)

Reply:
I remade it. I tried not to keep all words in memory, so decided to keep only 10 words in memory. So now it works without sorting and complexity is O(n*k), where k is 10 (constant).